Mlem/Views/Shared/Posts/Expanded Post.swift

```diff
@@ -72,25 +92,91 @@ struct ExpandedPost: View {
+                            if commentTracker.comments.isEmpty {
+                                noCommentsView()
+                            } else {
+                                commentsView
+                                    .onAppear {
+                                        if let target = scrollTarget {
+                                            scrollTarget = nil
+                                            scrollProxy.scrollTo(target, anchor: .top)
+                                        }
+                                    }
+                            }
+                        }
+                    }
+                    .onChange(of: scrollTarget) { target in
+                        if let target = target {
+                            scrollTarget = nil
+                            withAnimation {
+                                scrollProxy.scrollTo(target, anchor: .top)
+                            }
+                        }
+                    }
+                    .onPreferenceChange(AnchorsKey.self) { anchors in
+                        topVisibleCommentId = topCommentRow(of: anchors, in: proxy)
+                    }
                 }
+                .listStyle(PlainListStyle())
+            }
+            if showCommentJumpButton && commentTracker.comments.count > 1 {
+                JumpButtonView(onShortPress: scrollToNextComment, onLongPress: scrollToPreviousComment)
             }
         }
-        .listStyle(PlainListStyle())
         .fancyTabScrollCompatible()
     }
     
+    func scrollToNextComment() {
+        if let topVisibleId = topVisibleCommentId {
+            var nextComment: HierarchicalComment?
+            
+            if topVisibleId == 0 {
+                nextComment = commentTracker.commentsView.first
+            } else {
+                if let index = commentTracker.commentsView.firstIndex(where: {
+                    $0.commentView.comment.id == topVisibleId
+                }) {
+                    let slice = commentTracker.commentsView[(index+1)...]
+                    nextComment = slice.first(where: { $0.depth == 0 })
+                }
+            }
+            if let nextComment = nextComment {
+                scrollTarget = nextComment.commentView.comment.id
+            }
+        }
+    }
+    
+    func scrollToPreviousComment() {
+        if topVisibleCommentId == commentTracker.comments.first?.commentView.comment.id {
+            scrollTarget = 0
+        }
+        if let topVisibleCommentId = topVisibleCommentId {
+            if let index = commentTracker.commentsView.firstIndex(where: {
+                $0.commentView.comment.id == topVisibleCommentId
+            }) {
+                let slice = commentTracker.commentsView[..<index]
+                if let previousComment = slice.last(where: { $0.depth == 0 }) {
+                    scrollTarget = previousComment.commentView.comment.id
+                }
+            }
+        }
+    }
```

Comment:
All of this slicing and indexing is a lot of iteration to do what amounts to moving up and down a list. Would it be a total PITA to refactor this mechanic to something like:
- Store an array of top level comment IDs that updates when `commentTracker.comments` updates
- Store a map that maps each comment id to the index of the closest top-level comment id (this might require some slightly sneaky iteration but this and the previous step should be θ(n) with one loop)
- Store an index in that array that is the next comment to jump to
- Rework `topCommentRow` to use that map to return the index of the next item to jump to

This will let these two functions do practically no computation and just increment or decrement the index

If this is a ridiculous idea or infeasible for some reason I'm not seeing, please let me know--you wrote the code, I'm just reading it